lib/matplotlib/figure.py

```diff
@@ -1242,6 +1242,22 @@ def colorbar(
             fig.sca(current_ax)
             cax.grid(visible=False, which='both', axis='both')
 
+        if hasattr(mappable, "figure") and mappable.figure is not None:
+            # Get top level artists
+            mappable_host_fig = mappable.figure
+            self_host_fig = self.figure
+            if isinstance(mappable_host_fig, mpl.figure.SubFigure):
+                mappable_host_fig = mappable_host_fig.figure
+            if isinstance(self_host_fig, mpl.figure.SubFigure):
+                self_host_fig = self_host_fig.figure
```

Comment:
Given our discussion, it seems like this should get a test?

Reply:
Sure. These lines are for the subfigure cases. In the latest commit, there are tests regarding subfigures in `test_figure::test_warn_colorbar_mismatch`, line 1685:1702, which specifically include:
1. Test that no warning is emitted when calling fig.colorbar(subfig), where subfig is associated with the fig.
2. Test that no warning is emitted when calling subfig1.colorbar(subfig2), where both subfigs are associated with the same figure.
3. Test that a warning is emitted when calling subfig1.colorbar(subfig2), where subfig1 is not associated with the same figure as subfig2.

Which cases am I missing? Please let me know and I'll add them.

Reply:
...And the reason L1252 is not covered by test is that I did not test on subfig.colorbar(subfig).

Reply:
Oh, and the reason L1252 is not covered by test is that `self_host_fig` is always a figure object no matter if we call colorbar() on a subfigure or figure. I should remove L1251:1252 in the next commit. Thanks!